File: src/main/scala/fortress/solverinterface/SolverTemplate.java
package fortress.solverinterface;

import fortress.msfol.*;
import fortress.util.StopWatch;
import fortress.modelfind.*;

import java.io.Writer;
import java.io.IOException;

abstract class SolverTemplate implements SolverStrategy {
    
    @Override
    public ModelFinderResult solve(Theory theory, int timeoutMillis, Writer log) throws IOException {
        // template method
        
        log.write("Converting to solver format: ");
        log.flush();
        
        StopWatch conversionTimer = new StopWatch();
        conversionTimer.startFresh();
        
        convertTheory(theory, log);
        
        log.write(StopWatch.formatNano(conversionTimer.elapsedNano()) + "\n");
        log.flush();
        
        
        int remainingMillis = timeoutMillis - StopWatch.nanoToMillis(conversionTimer.elapsedNano());
        if(remainingMillis <= 0) {
            log.write("TIMEOUT within Fortress.\n");
            log.flush();
            return ModelFinderResult.Timeout();
        }
        
        updateTimeout(remainingMillis);
        
        log.write("Solving... ");
        log.flush();
        
        StopWatch solverTimer = new StopWatch();
        solverTimer.startFresh();
        
        ModelFinderResult result = runSolver(log);
        
        log.write("Z3 solver time: " + StopWatch.formatNano(solverTimer.elapsedNano()) + "\n");
        
        return result;
    }
    
    abstract protected void convertTheory(Theory theory, Writer log) throws IOException;
    abstract protected void updateTimeout(int remainingMillis);
    abstract protected ModelFinderResult runSolver(Writer log) throws IOException;
}

File: src/main/scala/fortress/solverinterface/SolverTemplate.scala
package fortress.solverinterface

import fortress.msfol._
import fortress.util._
import fortress.modelfind._

abstract class SolverTemplate extends SolverStrategy {
    
    override def solve(theory: Theory, timeoutMillis: Milliseconds, eventLoggers: Seq[EventLogger]): ModelFinderResult = {
        // template method
        
        for(logger <- eventLoggers) logger.convertingToSolverFormat()
        
        val conversionTimer = new StopWatch
        conversionTimer.startFresh()
        
        convertTheory(theory)
        
        for(logger <- eventLoggers) logger.convertedToSolverFormat(conversionTimer.elapsedNano())
        
        val remainingMillis = timeoutMillis - conversionTimer.elapsedNano().toMilli
        if(remainingMillis <= Milliseconds(0)) {
            for(logger <- eventLoggers) logger.timeoutInternal()
            return ModelFinderResult.Timeout
        }
        
        updateTimeout(remainingMillis)
        
        for(logger <- eventLoggers) logger.solving()
        
        val solverTimer = new StopWatch
        solverTimer.startFresh()
        
        val result: ModelFinderResult = runSolver()
        
        for(logger <- eventLoggers) logger.solverFinished(solverTimer.elapsedNano())
        logSMT2Output(eventLoggers)
        
        result
    }
    
    def logSMT2Output(eventLoggers: Seq[EventLogger]): Unit = { }
    
    protected def convertTheory(theory: Theory): Unit
    protected def updateTimeout(remainingMillis: Milliseconds): Unit
    protected def runSolver(): ModelFinderResult
}
